cmd/jumphost/create.go

```diff
@@ -0,0 +1,392 @@
+		Example: `
+  # Create and delete a jumphost
+  osdctl jumphost create --subnet-id public-subnet-id
+  osdctl jumphost delete --subnet-id public-subnet-id`,
+		Args: cobra.NoArgs,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			j, err := initJumphostConfig(context.TODO(), clusterId, subnetId)
+			if err != nil {
+				return err
+			}
+
+			return j.runCreate(context.TODO())
+		},
+	}
+
+	// create.Flags().StringVarP(&clusterId, "cluster-id", "c", "", "OCM internal/external cluster id trying to access via a jumphost")
+	create.Flags().StringVar(&subnetId, "subnet-id", "", "public subnet id to create a jumphost in")
+	create.MarkFlagRequired("subnet-id")
+
+	return create
+}
+
+func (j *jumphostConfig) runCreate(ctx context.Context) error {
+	if err := j.createKeyPair(ctx); err != nil {
+		return err
+	}
+
+	securityGroupId, err := j.createSecurityGroup(ctx)
+	if err != nil {
+		return err
+	}
+
+	if err := j.createEc2Jumphost(ctx, securityGroupId); err != nil {
+		return err
+	}
+
+	log.Println(j.assembleNextSteps())
+	return nil
+}
+
+// createKeyPair creates an EC2 key pair and attempts to save the private key to a temporary file and assign
+// 400 permissions. If it is able to do so successfully, it stores the filepath in j.keyFilePath.
+func (j *jumphostConfig) createKeyPair(ctx context.Context) error {
+	resp, err := j.awsClient.CreateKeyPair(ctx, &ec2.CreateKeyPairInput{
+		// TODO: Make name configurable if multiple SREs want to have an active jumphost
+		KeyName:   aws.String(awsResourceName),
+		KeyFormat: types.KeyFormatPem,
+		KeyType:   types.KeyTypeEd25519,
+		TagSpecifications: []types.TagSpecification{
+			{
+				ResourceType: types.ResourceTypeKeyPair,
+				Tags:         j.tags,
+			},
+		},
+	})
+	if err != nil {
+		return fmt.Errorf("failed to create keypair: %w", err)
+	}
+
+	f, err := os.CreateTemp("", "jumphost_*.pem")
+	if err != nil {
+		log.Printf("failed to create temp file: %s, printing private key instead", err)
+		log.Printf("created %s (%s). Save this private key\n%s", *resp.KeyName, *resp.KeyPairId, *resp.KeyMaterial)
+		return nil
+	}
+
+	defer f.Close()
+	if _, err := f.WriteString(*resp.KeyMaterial); err != nil {
+		log.Printf("failed to create temp file: %s, printing private key instead", err)
+		log.Printf("created %s (%s). Save this private key\n%s", *resp.KeyName, *resp.KeyPairId, *resp.KeyMaterial)
+		return nil
+	}
+
+	if err := f.Chmod(0400); err != nil {
+		log.Printf("failed to create temp file: %s, printing private key instead", err)
+		log.Printf("created %s (%s). Save this private key\n%s", *resp.KeyName, *resp.KeyPairId, *resp.KeyMaterial)
+		return nil
+	}
+
+	log.Printf("created file %s", f.Name())
+	j.keyFilepath = f.Name()
+	return nil
+}
+
+// createSecurityGroup creates a security group and creates a single inbound rule to allow the user's public IP to SSH.
+func (j jumphostConfig) createSecurityGroup(ctx context.Context) (string, error) {
```

Comment:
Should this method be on the pointer receiver?

Reply:
I have mixed them where it's necessary to modify `jumphostConfig` - in this case `createSecurityGroup` does not need to modify `jumphostConfig` so I made it a value receiver.

I'm ok with making them all pointer receivers so that in the future people don't try to mutate state with a value receiver and get unexpected results.

Reply:
Latest push converts all methods to act on to pointer receivers